airflow/models/dataset.py

```diff
@@ -66,10 +68,36 @@ def __init__(self, uri: str, **kwargs):
         super().__init__(uri=uri, **kwargs)
 
     def __eq__(self, other):
-        return self.uri == other.uri
+        if isinstance(other, self.__class__):
+            return self.uri == other.uri
+        raise NotImplemented
 
     def __hash__(self):
         return hash(self.uri)
 
     def __repr__(self):
         return f"{self.__class__.__name__}(uri={self.uri!r}, extra={self.extra!r})"
+
+    @provide_session
+    def get_dataset_id(self, session=NEW_SESSION):
+        if self.id:
+            dataset_id = self.id
+        else:
+            stored = session.query(self.__class__).filter(self.__class__.uri == self.uri).first()
```

Comment:
```suggestion
            stored = session.query(self.__class__.id).filter(self.__class__.uri == self.uri).first()
```

We can just bring back the id, no?